config/config.go

```diff
@@ -57,3 +58,161 @@ func FindAndReadConfig(fs afero.Fs, configFile string) (*v1.Config, error) {
 	c, err2 := ReadConfig(reader)
 	return c, err2
 }
+
+type ver struct {
+	Version int `json:"version"`
+}
+
+func detectVersion(b []byte) (int, error) {
+	v := &ver{}
+	err := json.Unmarshal(b, v)
+	if err != nil {
+		return 0, err
+	}
+	if v.Version == 0 {
+		return 1, nil
+	}
+	return v.Version, nil
+}
+
+// upgradeConfigVersion will convert a v1.Config to a v2.Config. Note that due to some semantic changes, this a lossy
+//  conversion
+func upgradeConfigVersion(c1 *v1.Config) (*v2.Config, error) {
+	c2 := &v2.Config{}
+
+	c2.Docker = c1.Docker
+	def1 := c1.Defaults
+	c2.Defaults = v2.Defaults{
+		Common: v2.Common{
+			Backend: v2.Backend{
+				Bucket:      def1.InfraBucket,
+				DynamoTable: def1.InfraDynamoTable,
+				Profile:     def1.AWSProfileBackend,
+				Region:      def1.AWSRegionBackend,
+			},
+			ExtraVars: def1.ExtraVars,
+			Providers: v2.Providers{
+				AWS: &v2.AWSProvider{
+					AccountID:         &def1.AccountID,
+					AdditionalRegions: def1.AWSRegions,
+					Profile:           &def1.AWSProfileProvider,
+					Region:            &def1.AWSRegionProvider,
+					Version:           &def1.TerraformVersion,
+				},
+			},
+			Owner:            def1.Owner,
+			Project:          def1.Project,
+			TerraformVersion: def1.TerraformVersion,
+		},
+	}
+
+	c2.Tools = v2.Tools{}
+	if def1.TfLint != nil {
+		c2.Tools.TfLint = def1.TfLint
+	}
+
+	if c1.TravisCI != nil {
+		c2.Tools.TravisCI = c1.TravisCI
+	}
+
+	c2.Accounts = map[string]v2.Account{}
+
+	var ptrstr = func(s *string) string {
+		if s == nil {
+			return ""
+		}
+		return *s
+	}
+
+	for acctName, acct := range c1.Accounts {
+		c2.Accounts[acctName] = v2.Account{
+			Common: v2.Common{
+				Backend: v2.Backend{
```

Comment:
If this will write the config back out to a file, might be worth ommiting these and instead let the inheritance chain take care of it

Reply:
I'm not sure I follow - if someone set these values at the account level (presumably overriding the defaults) then we should preserve them, yeah?

Reply:
Actually, not sure what the behavior of `omitempty` is on a zero struct and this might actually be good. Was thinking if none of `acct.InfraBucket`,`InfraDynamoTable`, etc are defined then we might not want `v2.acct.Backend` to show up in the serialized config we write back out to a file

Reply:
Good points. I think I just need to watch out for this when I write the code to migrate config files. I'm sure some tweaking will be required.